Add PORT_PHY_ATTR flex counter support

dhanasekar-arista opened the pull request:
Implement PORT_PHY_ATTR flex counter support for monitoring the port attributes (RX_SIGNAL_DETECT, FEC_ALIGNMENT_LOCK, RX_SNR).

### Changes
**meta/:** Add serialization support for PORT_SNR_LIST and PORT_LANE_LATCH_STATUS_LIST
**syncd/:** Introduce PortAttrContext extending AttrContext with specialized data handling for lane-based attribute types
**unittest/:** Add unit tests for serialization and collectData APIs.

### Implementation
Queries SAI for per-port lane count using BUFFER_OVERFLOW pattern
Maintains lane count cache for efficient memory allocation
Writes collected data to dedicated PORT_PHY_ATTR table in COUNTERS_DB

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

dhanasekar-arista committed:
Add PORT_PHY_ATTR flex counter support

This commit introduces support for polling PORT PHY attributes using the flex counter infrastructure.

Following changes are added in sairedis to support this feature
  - Added an extra parameter(AttrDataType) to the template of class AttrContext,
    this is required to handle attributes with complex data strucutres.
  - Introduce PortAttrContext extending AttrContext with
    specialized data handling for lane-based attribute types
  - Refactor serialization to use compact JSON dictionary format instead of
    verbose array-of-objects (e.g., {"0":"T*","1":"F"})
  - Support complex attribute value types (PORT_LANE_LATCH_STATUS_LIST,
    PORT_SNR_LIST) with proper memory management
  - Write collected metrics to PORT_PHY_ATTR_TABLE with attribute aliases:
    * phy_rx_signal_detect (latch status: T/F with * for changed)
    * pcs_fec_lane_alignment_lock (per-lane FEC lock status)
    * rx_snr (per-lane SNR values)
  - Add validation helpers and unit tests covering serialize/deserialize and
    full collection workflow with mocked SAI interface

Signed-off-by: dhanasekar-arista <dhanasekar@arista.com>

mssonicbld commented:
/azp run

dhanasekar-arista committed:
change port_attr to port_phy_attr

Signed-off-by: dhanasekar-arista <dhanasekar@arista.com>

dhanasekar-arista committed:
minor code refactor

Signed-off-by: dhanasekar-arista <dhanasekar@arista.com>

mssonicbld commented:
/azp run

prgeor commented on meta/SaiSerialize.cpp:
```diff
@@ -1621,45 +1628,56 @@ std::string sai_serialize_latch_status(
-        arr.push_back(item);
+        if (status_list.list[i].value.changed)
+        {
+            value += "*";
```
@dhanasekar-arista we need an entry in DB to know the time when the value changed because the next polling may overwrite this * value and SW may miss to there was a signal change?

dhanasekar-arista replied:
Please confirm the final format in which we want the output.
Shall we use this format?
{"0":["T*",1734960123456],"1":["F",1734960123456]}

I think millisec granularity should be more than sufficient and nanosec would be an overkill.

prgeor commented on syncd/FlexCounter.cpp:
```diff
@@ -7,11 +7,13 @@
 
 #include "FlexCounter.h"
 #include "VidManager.h"
+#include "SaiSwitch.h"
```
@dhanasekar-arista why this is needed?

prgeor commented on syncd/FlexCounter.cpp:
```diff
@@ -7,11 +7,13 @@
 
 #include "swss/redisapi.h"
 #include "swss/tokenize.h"
+#include "swss/schema.h"
```
@dhanasekar-arista why this is needed?

dhanasekar-arista replied:
not required will remove it.

prgeor commented on syncd/FlexCounter.cpp:
```diff
@@ -494,6 +498,15 @@ void deserializeAttr(
 }
 
+template <>
+void deserializeAttr(
```
@dhanasekar-arista should be renamed to `deserializePortAttr`?

dhanasekar-arista replied:
No, we cannot rename this. This is a template specialization of the generic deserializeAttr<AttrType> template function that's used by the AttrContext class template.

dhanasekar-arista committed:
Record timestamp and counter to track when Lane Latch status was updated

We need a timestamp and a counter to know if latch status  has changed between
successive queries, if yes how many time has it changed. Without this data we wont
know if it transitioned from T* -> F* -> T*.
So we will record this format of data for each lane,
<lane_number>:[<status>,<timestamp in ms since epoch>,<running_counter>]

Signed-off-by: dhanasekar-arista <dhanasekar@arista.com>

dhanasekar-arista commented:
@prgeor This change is ready for review. it cannot be cherry picked to 202505, hence ive created a separate PR https://github.com/sonic-net/sonic-sairedis/pull/1702.

prgeor commented on syncd/FlexCounter.cpp:
```diff
@@ -1699,6 +1721,427 @@ class AttrContext : public CounterContext<AttrType>
+     * @param attr_id Attribute ID (must be latch type)
+     * @param attr Current attribute containing lane latch status list
+     */
+    void updateLaneMetadata(
```
@dhanasekar-arista nit, if you can rename to `updateLatchedLaneMetadata`

prgeor commented on meta/SaiSerialize.cpp:
```diff
@@ -1052,6 +1052,13 @@ std::string sai_serialize_ipmc_entry_type(
     return sai_serialize_enum(type, &sai_metadata_enum_sai_ipmc_entry_type_t);
 }
 
+std::string sai_serialize_port_attr(_In_ const sai_port_attr_t port_attr)
```
@dhanasekar-arista I don't see this function being used/called?

dhanasekar-arista committed:
Fixed review comments

Signed-off-by: dhanasekar-arista <dhanasekar@arista.com>

dhanasekar-arista committed:
Rename portAttr to portPhyAttr

Signed-off-by: dhanasekar-arista <dhanasekar@arista.com>

dhanasekar-arista committed:
Bug Fix

Signed-off-by: dhanasekar-arista <dhanasekar@arista.com>

dhanasekar-arista commented:
@prgeor the builds for sonic-sairedis have been cancelled, can you please retrigger them?

And i frequently see the branch becomes out-of-date with the base, should i keep updating it often until it gets merged?

dhanasekar-arista committed:
Merge branch 'master' into port_serdes_flex_counter

dhanasekar-arista commented:
> > @prgeor the builds for sonic-sairedis have been cancelled, can you please retrigger them?
> > And i frequently see the branch becomes out-of-date with the base, should i keep updating it often until it gets merged?
> 
> @dhanasekar-arista please ignore the message showing out-of-date. don't trigger the build because of that.

okay, I'm not sure why the 2 checks are in cancelled state.

mssonicbld commented:
Cherry-pick PR to 202511: https://github.com/sonic-net/sonic-sairedis/pull/1776
